### DISCUSSION TITLE
Nombre d'axes indépendant ?

### DISCUSSION THREAD

**Discussion by StudentMaker01** [fra589/cn5X#47] [Q&A]
Bonjour,

Peut-on contrôler 6 axes indépendant sur base d'Arduino Mega ?
Quel est le firmware et/ou software le plus adapté (Windows 11)
Merci et bravo !

**Comment by fra589** [comment]
Bonjour @StudentMaker01,

J'ai écrit [cn5X](https://github.com/fra589/cn5X) en accord avec le firmware [grbl-Mega-5X](https://github.com/fra589/grbl-Mega-5X) dont j'assure la maintenance.
grbl-Mega-5X est capable de contrôler 6 axes indépendants sur Arduino Mega et cn5X est en mesure de gérer ces 6 axes.

@++;
Gauthier.

**Comment by StudentMaker01** [comment]
Bonjour,
Merci beaucoup pour votre réponse !
Comment paramètre-t-on le firmware ?
Il existe une documentation ?
Merci encore.
Serge.

Le mar. 26 nov. 2024, 19:08, Gauthier Brière ***@***.***> a
écrit :

**Reply by fra589** [reply]
Je tiens également la documentation de grbl-Mega-5X à jour sur le wiki :  
https://github.com/fra589/grbl-Mega-5X/wiki
Désolé, c'est rédigé en anglais :-)

S'il tu as des questions/besoin d’éclaircissements après lecture de la documentation, cherches dans les issues ou dans les discussions et ouvre une autre discussion si tu ne trouves pas.

En général, j'essaie de ne pas mettre trop de temps à répondre.

@++;
Gauthier.

**Comment by StudentMaker01** [comment]
Bonjour,
Je mets au point un projet pour l'enseignement, pourrait-on communiquer par
téléphone, j'aurais besoin de précisions avant de faire certains choix...et
votre aide pourrait être utile.
Cordialement.
Serge Febve.

Le mar. 26 nov. 2024, 19:37, Gauthier Brière ***@***.***> a
écrit :

**Comment by fra589** [comment]
Bonjour Serge,  
  
Comme la plupart des retraités, je manque cruellement de temps :-)
Je participe avec plaisir à la création et à la maintenance de différents logiciels et systèmes open source, mais je le fait uniquement quand je trouve un peu de temps par-ci/par-là... Toujours irrégulièrement, sans vouloir forcément tenir des objectifs et/ou des délais.
Je préfère donc utiliser le système de discussion (ou d'issues en cas de bugs) de Github, ce qui me permet d'y passer de temps en temps, d'y réfléchir en temps masqué (je n'ai pas toujours la réponse immédiate) et de répondre ensuite.

Je suis prêt à répondre à vos besoins de précisions, mais en continuant à utiliser le canal de discussions Github.

Bien cordialement,
Gauthier.

**Comment by StudentMaker01** [comment]
Bonjour,
je comprends.
En fait, nous avons fabriqué un bras robotisé et nous cherchons le
contrôleur et son firmware avec un logiciel de pilotage.
Nous sommes partis dans tous les sens mais il est difficile d'y voir clair
afin de trouver la solution.
Il y a beaucoup de choses sur Youtube, mais on ne dévoile jamais l'ensemble
des éléments. il manque toujours quelque chose !
Si vous avez une idée, elle est la bienvenue !
Cordialement.
Serge.

Le ven. 29 nov. 2024 à 18:00, Gauthier Brière ***@***.***> a
écrit :

**Comment by fra589** [comment]
Quels moteurs utilise votre bras robotisé ?

**Comment by StudentMaker01** [comment]
Des moteurs pas à pas classiques NEMA 17, NEMA 11.

Le ven. 29 nov. 2024 à 19:04, Gauthier Brière ***@***.***> a
écrit :

**Comment by fra589** [comment]
Les moteurs sont donc assez facile à piloter avec une "petite électronique" :-)

A l'origine, j'ai transformé le grbl-Mega d'origine (limité à 3 axes) pour le porter à 5, puis 6 axes tout en conservant l'électronique standard (et très peu cher) des imprimantes 3D de l'époque : Arduino Mega 2560 + RAMPS + drivers Pololu A4988 ou DRV8834.

Pour utiliser le 6ème axe, il faut ajouter à la RAMP une petite carte comme par exemple celle ci : 
https://fluxworkshop.com/products/bmaa100085-9v-a4988-controller-inc-driver-keyes-red?variant=32458611884092
cf. https://www.google.com/search?q=keyes+stepper
Ou bien on peux également se passer de la carte RAMPS et utiliser des drivers pas à pas indépendant directement connectés à la carte Arduino.

Mon but initial était de construire ma propre fraiseuse CNC à base des matériels peu cher utilisées sur les imprimantes 3D open source.

Ensuite, ne trouvant pas de logiciel de pilotage compatible 5 axes j'ai décidé d'écrire le mien : cn5X.

cn5X permet de piloter manuellement les 6 axes en mode "jog" (boutons de mouvements) ou en entrant les commande GCode dans l'interface.
Il permet également de charger des fichiers de programmes complet et de les faire exécuter par la machine.

Finalement, la chaine de pilotage de la machine se présente comme ça :
![Chaine CFAO](https://github.com/user-attachments/assets/39198eae-f185-4af5-be97-7b4d7d7461ca)

@++;
Gauthier.

**Comment by StudentMaker01** [comment]
Oui c'est exactement la plateforme matériel que j'ai mais peut-être que
pour commander un bras ce n'est pas adapté ?

Le ven. 29 nov. 2024, 20:05, Gauthier Brière ***@***.***> a
écrit :

**Comment by fra589** [comment]
il n'y a pas de raison que ça ne fonctionne pas...
Dans grbl-Mega-5X, on peu configurer tous les axes comme étant linéaires ou rotatif.
les unités utilisées sont configurables à l'aide de paramètres qui font correspondre un nombre d'unité de mouvement par rapport au nombre de pas moteurs correspondants.

**Comment by StudentMaker01** [comment]
Bonjour,
grbl-Mega-5X est maintenant compilé et installé sur l'arduino.
J'ai modifié les axes rotatifs à 6 et linéaires à 0.
Malheureusement cn5X ne se lance pas.
La fenêtre apparaît à peine et disparaît immédiatement.
Je suis sous Win11 j'ai Python installé, je ne vois pas ce qui peut bloquer.
Merci.

Le ven. 29 nov. 2024 à 20:15, Gauthier Brière ***@***.***> a
écrit :

**Comment by fra589** [comment]
Bonjour Serge,

Pour que nous puissions voir ce qui se passe, il faut que tu utilises l'invite de commande pour lancer cn5X.

Je n'ai pas Windows 11 chez moi (il reste juste un vieux Windows 10 parmi tous mes ordinateurs Linux) donc je ne sais pas si la procédure pour lancer l'invite de commande est toujours la même...
- Appuyer sur les touches `<Windows> + s` pour lancer la recherche Windows,
- Taper `cmd` dans la zone de recherche,
- "Invite de commande" devrait apparaitre comme meilleur résultat,
- Double-Cliquer dessus pour l'exécuter.

Dans la fenêtre invite de commande, tapez : 
```
> cd <chemin\ou\est\installé\cn5X>
> cn5X.px

```

Le programme cn5X devrait démarrer, et en cas d'erreur, afficher des messages...

Merci de transmettre ces messages.

@++;
Gauthier.

**Comment by StudentMaker01** [comment]
Bonjour,
J'ai déplacé le dossier à la racine du système pour éviter les chemins à
rallonge.
Capture d'écran du message.
Merci !

Le mar. 3 déc. 2024 à 07:37, Gauthier Brière ***@***.***> a
écrit :

**Reply by fra589** [reply]
Je ne vois pas la capture ?

**Comment by StudentMaker01** [comment]
Message :
Traceback (most recent call last):
  File "C:\cn5x\cn5X.py", line 29, in <module>
    import argparse
ImportError: No module named argparse

Le mar. 3 déc. 2024 à 08:17, Gauthier Brière ***@***.***> a
écrit :

**Comment by StudentMaker01** [comment]
il manque "argparse".

**Comment by fra589** [comment]
Tu peux me donner le résultat de la commande : 
`> python --version`

**Comment by fra589** [comment]
ainsi que le résultat de 
`> pip list`

**Comment by StudentMaker01** [comment]
Python introuvable
pip n'est pas reconnu en temps que commande interne.

Problème d'installation de Python ?

Le mar. 3 déc. 2024 à 08:25, Gauthier Brière ***@***.***> a
écrit :

**Reply by fra589** [reply]
> pip n'est pas reconnu en temps que commande interne.

Si tu ne t'étais pas rendu compte que pip ne fonctionne pas, tu n'as probablement pas effectué l'installation complette pour que cn5X fonctionne...
C'est documenté ici : https://github.com/fra589/cn5X/wiki/Install-cn5X-on-Windows

- Installation de Python 3 depuis python.org avec ajout de Python dans le PATH,
- `> pip install pyserial`
-  `> pip install pyqt6`

**Comment by fra589** [comment]
Effectivement, tu as probablement un problème d'installation de Python ou de PATH (chemin de recherche des programmes)

As-tu bien ajouté Python au PATH et associé l'extension de fichier ".py" ? (ce sont des options lors de l'installation de python)

**Comment by fra589** [comment]
Chez moi : 
```
C:\Users\Gauthier>python -c "import argparse; print (argparse)"
<module 'argparse' from 'C:\\Program Files\\Python311\\Lib\\argparse.py'>
```
Python est installé dans `C:\Program Files\Python311`

**Comment by fra589** [comment]
L'installation de Python : 

![installPython01](https://github.com/user-attachments/assets/1fa78480-0549-4bce-9f54-744012aee661)

![installPython03](https://github.com/user-attachments/assets/b5763d97-f848-4691-90b8-69cf2bca54ca)

![installPython04](https://github.com/user-attachments/assets/81e16279-c16e-4c9a-8d24-a1c055e79b0c)

![installPython05](https://github.com/user-attachments/assets/1d4b2a11-55c4-42b7-93fc-e9217c3cb332)

**Comment by StudentMaker01** [comment]
je viens tout nettoyer et de ré-installer tout ça...ces trucs de
développeurs, c'est intimidant quand on y connait rien !
cn5X vient de s'ouvrir pour la première fois !
Merci beaucoup.

Le mar. 3 déc. 2024 à 08:46, Gauthier Brière ***@***.***> a
écrit :

**Reply by fra589** [reply]
je vais regarder si je peux faire un exe installable/utilisable sous Windows sans avoir à installer Python et sa clique...
Mais faut pas être trop pressé :-D

**Comment by StudentMaker01** [comment]
Connection à l'arduino MEGA, un bip se fait entendre, grbl settings
OverFlowError et la fenêtre se ferme...

**Reply by fra589** [reply]
Bonjour Serge,  
  
Tu ne m'as pas envoyé les messages d'erreur... 
As tu réussi à faire fonctionner cn5X et grbl-Mega-5X ?
Si oui, quelle était l'erreur ?

@++;
Gauthier

**Comment by fra589** [comment]
Hummm...  
Tu peux refaire à partir de l'invite de commande et envoyer tous les message en sortie ?

**Comment by StudentMaker01** [comment]
Bonjour,
Juste le temps de voir dans la fenêtre "MS DOS" en arrière plan
"OverFlowError" et tout se ferme...

Le jeu. 12 déc. 2024, 08:03, Gauthier Brière ***@***.***> a
écrit :

**Reply by fra589** [reply]
Bonjour Serge,

Pour avoir le temps de lire le message d'erreur dans la fenêtre "Invite de commande", il faut que tu refasse comme je te l'avais indiqué lors de l'erreur précédente.
 Lancer l'invite de commande :
- Appuyer sur les touches `<Windows> + s` pour lancer la recherche Windows,
- Taper `cmd` dans la zone de recherche,
- "Invite de commande" devrait apparaitre comme meilleur résultat,
- Double-Cliquer dessus pour l'exécuter.

Dans la fenêtre invite de commande, tapez : 
```
> cd <chemin\ou\est\installé\cn5X>
> cn5X.px

```

Le programme cn5X devrait démarrer, au moment de l'erreur, afficher des messages...

M'envoyer ces messages sont importants pour moi, car ils me permettent comprendre ce qui s'est passé et de corriger en cas de bug avéré...

@++;
Gauthier.

**Comment by StudentMaker01** [comment]
J'avoue que ça m'a un peu découragé, j'ai vu également beaucoup de démos
d'un soft RoboDK, ça a l'air intéressant mais pareil...la reconnaissance et
la communication avec une plateforme Arduino sont mal ou pas du tout
décrites...et ça reste un "mystères" entre initiés...
Je suis assez dubitatif quand à la poursuite du projet...

**Comment by StudentMaker01** [comment]
Lien vers une vidéo de ce dont je rêve :
https://youtu.be/fNdqvi81djM?si=VQqYyvu0cwNUx_j2

**Reply by fra589** [reply]
Oui, a ce prix là, ça peut (doit) être beau !

A quasiment 4000€ la licence plus 1500€ par an à partir de la deuxième année, tu dois pouvoir les contacter pour qu'ils t'aident à réaliser ton rêve !

@++;
Gauthier.

**Comment by StudentMaker01** [comment]
Bonjour,
N'étant pas chez moi pour le moment je ne pourrai pas avoir d'autre vue sur
le message mais il me semble bien que ce soit tout ce qui s'est affiché.
Oui RoboDK est hors de prix !
Mais dans l'idée s'il était possible d'avoir un environnement et une
interface similaire, ce serait quand même pas mal du tout... visuellement
c'est un "avantage confortable"...
Depuis avec un ami, nous avons développé un code Arduino et une application
Windows, ça fonctionne.
Le souci est que c'est donc un ensemble cohérent mais pas universel.
J'aurais préféré avoir la possibilité d'utiliser le classique GCODE.
C'est répandu, relativement intuitif et il existe des dizaines de softs
permettant d'envoyer un programme ou un ensemble de commandes.
Cela serait beaucoup plus ouvert et plus général. L'idée d'avoir une
application dédiée, j'ai peur que ce soit "limitant"...

Le ven. 13 déc. 2024, 07:07, Gauthier Brière ***@***.***> a
écrit :
